File: code/3502/3502不同路径数.cpp
#include <iostream>
#include <cstring>
#include <algorithm>
#include <set>
using namespace std;

const int N = 10;
int dx[] = {-1, 0, 1, 0}, dy[] = {0, -1, 0, 1};
int g[N][N];
int n, m, k;
set<int> res;

void dfs(int x, int y, int u, int cur) {
	if (u > k) {
		res.insert(cur);
		return;
	}
	cur = 10 * cur + g[x][y];
	for (int i = 0; i < 4; i++) {
		int a = x + dx[i], b = y + dy[i];
		if (a < 0 || a >= n || b < 0 || b >= m)
			continue;
		dfs(a, b, u + 1, cur);
	}
}

int main() {
	cin >> n >> m >> k;
	for (int i = 0; i < n; i++)
		for (int j = 0; j < m; j++)
			cin >> g[i][j];

	for (int i = 0; i < n; i++)
		for (int j = 0; j < m; j++)
			dfs(i, j, 0, 0);

	cout << res.size() << endl;

	return 0;
}

File: code/3502/3502不同路径数.py
def dfs(x, y, u, cur):
    if u > k:
        res.add(cur)
        return

    dx, dy = [-1, 0, 1, 0], [0, -1, 0, 1]
    for i in range(4):
        a, b = x + dx[i], y + dy[i]
        if a < 0 or a >= n or b < 0 or b >= m:
            continue
        dfs(a, b, u + 1, 10 * cur + g[x][y])


if __name__ == "__main__":
    n, m, k = map(int, input().split())
    g = [list(map(int, input().split())) for _ in range(n)]
    res = set()
    for i in range(n):
        for j in range(m):
            dfs(i, j, 0, 0)
    print(len(res))
